heat/core/statistics.py

```diff
@@ -626,77 +627,94 @@ def maximum(x1, x2, out=None):
     >>> ht.maximum(a, d)
     ValueError: operands could not be broadcast, input shapes (3, 4) (3, 4, 5)
     """
-    # perform sanitation
-    if not isinstance(x1, dndarray.DNDarray) or not isinstance(x2, dndarray.DNDarray):
-        raise TypeError(
-            "expected x1 and x2 to be a ht.DNDarray, but were {}, {} ".format(type(x1), type(x2))
-        )
-    if out is not None and not isinstance(out, dndarray.DNDarray):
-        raise TypeError("expected out to be None or an ht.DNDarray, but was {}".format(type(out)))
-
-    # apply split semantics
-    if x1.split is not None or x2.split is not None:
-        if x1.split is None:
-            x1.resplit_(x2.split)
-        if x2.split is None:
-            x2.resplit_(x1.split)
-        if x1.split != x2.split:
-            if np.prod(x1.gshape) < np.prod(x2.gshape):
-                x1.resplit_(x2.split)
-            if np.prod(x2.gshape) < np.prod(x1.gshape):
-                x2.resplit_(x1.split)
+    # TODO: move all of this to __binary_op
+    # sanitize input
+    in_x = [x1, x2]
+    splits = []
+    shapes = []
+    comms = []
+    devices = []
+    t_x = []
+    for i, x in enumerate(in_x):
+        if isinstance(x, (int, float)):
+            x = factories.array([x], dtype=types.canonical_heat_type(type(x)), split=None)
+            in_x[i] = x
+        if not isinstance(x, dndarray.DNDarray):
+            raise TypeError("expected two DNDarrays, but array {} was {} ".format(i, type(x)))
+        if x.size == 0:
+            raise ValueError("operation not possible, array {} contains no data".format(i))
+        splits.append(x.split)
+        shapes.append(x.gshape)
+        comms.append(x.comm)
+        devices.append(x.device.device_type)
+        t_x.append(x._DNDarray__array)
+
+    # sanitize devices
+    if devices.count("gpu") == 1:
+        not_on_gpu = abs(devices.index("gpu") - 1)
+        # move to gpu
+        in_x[not_on_gpu] = in_x[not_on_gpu].gpu()
```

Comment:
@mtar do you think this check is needed? I'm thinking it might be redundant with the new setup

Reply:
I don't think you will need it if an error is raised on different devices. However, you will need to specify the device in line 640 explicitly.

Reply:
I see. Hm, but at line 640 I don't know yet on what device the other array is. 

I'll leave this check in. It's just that it can't be tested  and codecov complains. I guess we'll have to live with that. Thanks @mtar!